File: CommonUtils/src/com/rarnu/utils/UIUtils.java
package com.rarnu.utils;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.PointF;
import android.graphics.Rect;
import android.util.DisplayMetrics;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.GridView;
import android.widget.ListView;

public class UIUtils {

    public static final int ACTIONBAR_HEIGHT = android.R.attr.actionBarSize;
    private static Context context = null;
    private static DisplayMetrics dm = null;
    private static boolean followSystemBackground = true;

    public static DisplayMetrics getDM() {
        return dm;
    }

    public static float getDensity() {
        return dm.density;
    }

    public static void initDisplayMetrics(Context ctx, WindowManager wm, boolean isFollowSystemBackground) {
        if (context == null) {
            context = ctx;
        }

        dm = new DisplayMetrics();
        wm.getDefaultDisplay().getMetrics(dm);
        followSystemBackground = isFollowSystemBackground;

    }

    public static boolean touchInDialog(Activity activity, MotionEvent e) {
        int leftW, rightW, topH, bottomH;
        leftW = 8;
        rightW = dm.widthPixels - leftW;
        topH = 0;
        bottomH = 450;
        return ((e.getX() > leftW) && (e.getX() < rightW) && (e.getY() > topH) && (e.getY() < bottomH));
    }

    public static boolean isScreenCenter(MotionEvent e) {
        boolean ret = true;
        if (e.getX() < (dm.widthPixels / 2 - 25)) {
            ret = false;
        }
        if (e.getX() > (dm.widthPixels / 2 + 25)) {
            ret = false;
        }
        if (e.getY() < (dm.heightPixels / 2 - 25)) {
            ret = false;
        }
        if (e.getY() > (dm.heightPixels / 2 + 25)) {
            ret = false;
        }
        return ret;
    }

    public static PointF getLeftBottomPoint() {
        return new PointF((dm.widthPixels / 4) + 0.09f, (dm.heightPixels / 4 * 3) + 0.09f);
    }

    public static PointF getRightBottomPoint() {
        return new PointF((dm.widthPixels / 4 * 3) + 0.09f, (dm.heightPixels / 4 * 3) + 0.09f);
    }

    public static PointF getLeftPoint() {
        return new PointF(20, dm.heightPixels / 2);
    }

    public static PointF getRightPoint() {
        return new PointF(dm.widthPixels - 20, dm.heightPixels / 2);
    }

    public static boolean isTouchLeft(MotionEvent e) {
        return (e.getX() < (dm.widthPixels / 2));
    }

    public static int getStatusbarHeight(Activity activity) {
        int statusHeight = 0;
        Rect localRect = new Rect();
        activity.getWindow().getDecorView().getWindowVisibleDisplayFrame(localRect);
        statusHeight = localRect.top;
        if (statusHeight == 0) {
            try {
                Class<?> localClass = Class.forName("com.android.internal.R$dimen");
                Object localObject = localClass.newInstance();
                int i5 = Integer.parseInt(localClass.getField("status_bar_height").get(localObject).toString());
                statusHeight = activity.getResources().getDimensionPixelSize(i5);
            } catch (Exception e) {

            }
        }
        return statusHeight;
    }

    public static void setActivitySizePos(Activity activity, int x, int y, int width, int height) {
        WindowManager.LayoutParams p = activity.getWindow().getAttributes();
        p.x = x;
        p.y = y;
        p.width = width;
        p.height = height;
        activity.getWindow().setAttributes(p);
    }

    public static int dipToPx(int dip) {
        if (dm == null) {
            return -1;
        }
        return (int) (dip * dm.density + 0.5f);
    }

    public static float pxToScaledPx(int px) {
        if (dm == null) {
            return -1;
        }
        return px / dm.density;
    }

    public static int scaledPxToPx(float scaledPx) {
        if (dm == null) {
            return -1;
        }
        return (int) (scaledPx * dm.density);
    }

    public static int countViewAdvWidth(int count, int innerMargin, int outerMargin) {
        int width = dm.widthPixels - (outerMargin * 2);
        width = width - (innerMargin * (count - 1));
        width = width / count;
        return width;
    }

    public static int countViewAdvWidthByFrame(int count, int innerMargin, int outerMargin, int frameWidth) {
        int width = frameWidth - (outerMargin * 2);
        width = width - (innerMargin * (count - 1));
        width = width / count;
        return width;
    }

    public static int countViewAdvHeight(int count, int innerMargin, int outerMargin) {
        int height = dm.heightPixels - (outerMargin * 2);
        height = height - (innerMargin * (count - 1));
        height = height / count;
        return height;
    }

    public static int countViewAdvHeightByFrame(int count, int innerMargin, int outerMargin, int frameHeight) {
        int height = frameHeight - (outerMargin * 2);
        height = height - (innerMargin * (count - 1));
        height = height / count;
        return height;
    }

    public static int getWidth() {
        return dm.widthPixels;
    }

    public static int getHeight() {
        return dm.heightPixels;
    }

    public static void setActivitySizeX(Activity a, int size) {
        WindowManager.LayoutParams lp = a.getWindow().getAttributes();
        lp.width = size;
        a.getWindow().setAttributes(lp);
    }

    public static void setActivitySizeY(Activity a, int size) {
        WindowManager.LayoutParams lp = a.getWindow().getAttributes();
        lp.height = size;
        a.getWindow().setAttributes(lp);
    }

    public static void setActivityPercentX(Activity a, float percent) {
        WindowManager.LayoutParams lp = a.getWindow().getAttributes();
        lp.width = (int) (getWidth() * percent / 100);
        a.getWindow().setAttributes(lp);
    }

    public static void setActivityPercentY(Activity a, float percent) {
        WindowManager.LayoutParams lp = a.getWindow().getAttributes();
        lp.height = (int) (getHeight() * percent / 100);
        a.getWindow().setAttributes(lp);
    }

    public static void setViewSizeX(View v, int size) {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        lp.width = size;
        v.setLayoutParams(lp);
    }

    public static void setViewSizeY(View v, int size) {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        lp.height = size;
        v.setLayoutParams(lp);
    }

    public static void setViewPercentX(View v, float percent) {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        lp.width = (int) (getWidth() * percent / 100);
        v.setLayoutParams(lp);
    }

    public static void setViewPercentY(View v, float percent) {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        lp.height = (int) (getHeight() * percent / 100);
        v.setLayoutParams(lp);
    }

    public static void setViewPercentXByFrame(View v, float percent, float frameXPercent) {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        lp.width = (int) ((getWidth() * frameXPercent / 100) * percent / 100);
        v.setLayoutParams(lp);
    }

    public static void setViewPercentYByFrame(View v, float percent, float frameYPercent) {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        lp.height = (int) ((getHeight() * frameYPercent / 100) * percent / 100);
        v.setLayoutParams(lp);
    }

    public static void setViewMarginPercent(View v, float marginLeft, float marginTop, float marginRight, float marginBottom)
            throws Exception {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        if (lp instanceof ViewGroup.MarginLayoutParams) {
            ViewGroup.MarginLayoutParams mlp = (ViewGroup.MarginLayoutParams) lp;
            mlp.leftMargin = (int) (getWidth() * marginLeft / 100);
            mlp.topMargin = (int) (getHeight() * marginTop / 100);
            mlp.rightMargin = (int) (getWidth() * marginRight / 100);
            mlp.bottomMargin = (int) (getHeight() * marginBottom / 100);
            v.setLayoutParams(mlp);
        } else {
            throw new Exception(context.getString(R.string.exception_not_margin_layout_param));
        }
    }

    public static void setViewMarginSize(View v, int marginLeft, int marginTop, int marginRight, int marginBottom) throws Exception {
        ViewGroup.LayoutParams lp = v.getLayoutParams();
        if (lp instanceof ViewGroup.MarginLayoutParams) {
            ViewGroup.MarginLayoutParams mlp = (ViewGroup.MarginLayoutParams) lp;
            mlp.leftMargin = marginLeft;
            mlp.topMargin = marginTop;
            mlp.rightMargin = marginRight;
            mlp.bottomMargin = marginBottom;
            v.setLayoutParams(mlp);
        } else {
            throw new Exception(context.getString(R.string.exception_not_margin_layout_param));
        }
    }

    public static void setViewPaddingPercent(View v, float paddingLeft, float paddingTop, float paddingRight, float paddingBottom) {
        int pLeft = (int) (getWidth() * paddingLeft / 100);
        int pTop = (int) (getHeight() * paddingTop / 100);
        int pRight = (int) (getWidth() * paddingRight / 100);
        int pBottom = (int) (getHeight() * paddingBottom / 100);
        v.setPadding(pLeft, pTop, pRight, pBottom);
    }

    public static void makeListViewFullSize(ListView lv, int itemHeight) {
        int itemCount = lv.getAdapter().getCount();
        int divider = lv.getDividerHeight();
        int height = (itemHeight + divider) * itemCount;
        ViewGroup.LayoutParams lp = lv.getLayoutParams();
        lp.height = height;
        lv.setLayoutParams(lp);
    }

    public static void makeGridViewFullSize(GridView gv, int itemHeight, int rowNum) {
        int itemCount = gv.getAdapter().getCount();
        int lines = (int) (itemCount / rowNum);
        if (itemCount % rowNum != 0) {
            lines++;
        }
        ViewGroup.LayoutParams lp = gv.getLayoutParams();
        lp.height = lines * itemHeight;
        gv.setLayoutParams(lp);

    }

    public static int getActionBarHeight() {
        TypedArray a = context.obtainStyledAttributes(new int[]{ACTIONBAR_HEIGHT});
        int ret = a.getDimensionPixelSize(0, -1);
        a.recycle();
        return ret;
    }

    public static boolean isFollowSystemBackground() {
        return followSystemBackground;
    }

    public static void setFollowSystemBackground(boolean isFollowSystemBackground) {
        followSystemBackground = isFollowSystemBackground;
    }
}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/utils/UIUtils.kt
package com.rarnu.tools.neo.utils

import android.app.Activity
import android.app.ActivityOptions
import android.content.Context
import android.os.Build
import android.util.DisplayMetrics
import android.view.*
import android.widget.GridView
import android.widget.ListView
import android.widget.SearchView

/**
 * Created by rarnu on 3/23/16.
 */
object UIUtils {

    val ACTIONBAR_HEIGHT = android.R.attr.actionBarSize
    private var context: Context? = null
    var dm: DisplayMetrics? = null
    var isFollowSystemBackground = false
    var density: Float? = null
        get() {
            return dm?.density
        }
    var width: Int? = null
        get() {
            return dm?.widthPixels
        }
    var height: Int? = null
        get() {
            return dm?.heightPixels
        }

    fun initDisplayMetrics(ctx: Context, wm: WindowManager, follow: Boolean) {
        context = ctx
        dm = DisplayMetrics()
        wm.defaultDisplay.getMetrics(dm)
        isFollowSystemBackground = follow
    }

    fun dip2px(dip: Int): Int {
        if (dm == null) {
            return -1
        }
        return (dip * dm!!.density + 0.5f).toInt()
    }

    fun px2scaled(px: Int): Float {
        if (dm == null) {
            return -1.0f
        }
        return px / dm!!.density
    }

    fun scaled2px(scaled: Float): Int {
        if (dm == null) {
            return -1
        }
        return (scaled * dm!!.density).toInt()
    }

    var actionBarHeight: Int? = null
        get() {
            val a = context?.obtainStyledAttributes(intArrayOf(ACTIONBAR_HEIGHT))
            val ret = a?.getDimensionPixelSize(0, -1)
            a?.recycle()
            return ret
        }

    var statusBarHeight: Int?= null
        get() {
            val resId = context?.resources?.getIdentifier("status_bar_height", "dimen", "android")
            val height = context?.resources?.getDimensionPixelSize(resId!!)
            return height
        }

    var navigationBarHeight: Int? = null
        get() {
            val resId = context?.resources?.getIdentifier("navigation_bar_height", "dimen", "android")
            val height = context?.resources?.getDimensionPixelSize(resId!!)
            return height
        }

    var hasNavigationBar = false
        get() {
            if (Build.MODEL.toLowerCase().contains("nexus")) {
                val hasMenuKey = ViewConfiguration.get(context).hasPermanentMenuKey()
                val hasBackKey = KeyCharacterMap.deviceHasKey(KeyEvent.KEYCODE_BACK)
                return (!hasMenuKey && !hasBackKey)
            } else {
                return false
            }
        }

    fun makeListViewFullSize(lv: ListView, itemHeight: Int) {
        val itemCount = lv.adapter.count
        val divider = lv.dividerHeight
        val height = (itemHeight + divider) * itemCount
        lv.layoutParams.height = height
    }

    fun makeGridViewFullSize(gv: GridView, itemHeight: Int, colNum: Int) {
        val itemCount = gv.adapter.count
        val lines = (itemCount / colNum).toInt() + (if (itemCount % colNum != 0) { 1 } else { 0 })
        gv.layoutParams.height = itemHeight * lines
    }

    fun setActivitySizePos(activity: Activity, x: Int, y: Int, width: Int, height: Int) {
        with(activity.window.attributes) {
            this.x = x
            this.y = y
            this.width = width
            this.height = height
        }
    }

    fun setActivityWidth(activity: Activity, width: Int) {
        activity.window.attributes.width = width
    }

    fun setActivityHeight(activity: Activity, height: Int) {
        activity.window.attributes.height = height
    }

    fun setActivityWidthPercent(activity: Activity, percent: Float) {
        activity.window.attributes.width = (width!! * percent / 100.0f).toInt()
    }

    fun setActivityHeightPercent(activity: Activity, percent: Float) {
        activity.window.attributes.height = (height!! * percent / 100.0f).toInt()
    }

    fun setViewWidth(v: View, width: Int) {
        v.layoutParams.width = width
    }

    fun setViewHeight(v: View, height: Int) {
        v.layoutParams.height = height
    }

    fun setViewWidthPercent(v: View, percent: Float) {
        v.layoutParams.width = (width!! * percent / 100.0f).toInt()
    }


    fun setViewHeightPercent(v: View, percent: Float) {
        v.layoutParams.height = (height!! * percent / 100.0f).toInt()
    }

    fun setViewMargin(v: View, left: Int, top: Int, right: Int, bottom: Int) {
        if (v.layoutParams is ViewGroup.MarginLayoutParams) {
            with(v.layoutParams as ViewGroup.MarginLayoutParams) {
                leftMargin = left
                topMargin = top
                rightMargin = right
                bottomMargin = bottom
            }
        }
    }

    fun setViewMarginPercent(v: View, left: Int, top: Int, right: Int, bottom: Int) {
        if (v.layoutParams is ViewGroup.MarginLayoutParams) {
            with(v.layoutParams as ViewGroup.MarginLayoutParams) {
                leftMargin = (UIUtils.width!! * left / 100.0f).toInt()
                topMargin = (UIUtils.height!! * top / 100.0f).toInt()
                rightMargin = (UIUtils.width!! * right / 100.0f).toInt()
                bottomMargin = (UIUtils.height!! * bottom / 100.0f).toInt()
            }
        }
    }

    fun setImmersion(activity: Activity, immersion: Boolean) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (immersion) {
                with(activity.window) {
                    addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
                    if (hasNavigationBar) {
                        addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
                    }
                }
            }
        }
    }

    fun setFitSystem(activity: Activity, res: Int, immersion: Boolean) {
        val v = activity.findViewById(res)
        if (v is ViewGroup) {
            setFitSystem(v, immersion)
        }
    }

    fun setFitSystem(v: ViewGroup, immersion: Boolean) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (immersion) {
                v.fitsSystemWindows = true
                v.setClipToPadding(true)
            }
        }
    }

    fun setSearchViewTextBackground(sv: SearchView, backgroundRes: Int) = try {
        val clz = sv.javaClass
        val fSearchPlate = clz.getDeclaredField("mSearchPlate")
        fSearchPlate.isAccessible =  true
        val vSearchPlate = fSearchPlate.get(sv) as View
        vSearchPlate.setBackgroundResource(backgroundRes)
        val fSubmitArea = clz.getDeclaredField("submit_area")
        fSubmitArea.isAccessible = true
        val vSubmitArea = fSubmitArea.get(sv) as View
        vSubmitArea.setBackgroundResource(backgroundRes)
    } catch(e: Exception) {

    }

}